File: problems/scaler/sort/quick_sort.js
/**
 * Q3. Quick Sort
 *
 * Given an integer array A, sort the array using Quick Sort.
 */

function soln(A) {
  function partition(arr, s, e) {
    let pivot = s;
    const mid = parseInt((s + e) / 2);
    if (
      (arr[s] <= arr[mid] && arr[s] >= arr[e]) ||
      (arr[s] <= arr[e] && arr[s] >= arr[mid])
    ) {
      pivot = s;
    } else if (
      (arr[mid] <= arr[s] && arr[mid] >= arr[e]) ||
      (arr[mid] <= arr[e] && arr[mid] >= arr[s])
    ) {
      pivot = mid;
    } else {
      pivot = e;
    }
    [arr[s], arr[pivot]] = [arr[pivot], arr[s]];
    pivot = s;
    let l = pivot + 1;
    let r = e;
    while (l <= r) {
      if (arr[l] < arr[pivot]) {
        l++;
      } else if (arr[r] > arr[pivot]) {
        r--;
      } else {
        [arr[l], arr[r]] = [arr[r], arr[l]];
        l++;
        r--;
      }
    }
    [arr[pivot], arr[r]] = [arr[r], arr[pivot]];

    return r;
  }

  function quickSort(arr, s, e) {
    if (s >= e) {
      return;
    }

    const pivot = partition(arr, s, e);
    quickSort(arr, s, pivot - 1);
    quickSort(arr, pivot + 1, e);
  }

  quickSort(A, 0, A.length - 1);
  return A;
}

console.log(
  soln([
    -58, 32, 10, 42, 82, -12, 88, 17, 35, 8, -72, -32, -48, 76, -48, -8, -69,
    -1, 54, 52, 37, -1, 95, 0, -32, 38, 62, 68, 68, -71, 62, -58, -25, 55, -11,
    2,
  ])
);

File: src/main/java/com/dsa/problems/scaler/sort/quick_sort.java
package com.dsa.problems.scaler.sort;

import java.util.ArrayList;
import java.util.Arrays;

public class quick_sort {
  private static int parition(ArrayList<Integer> A, int s, int e) {
    int m = (int)((s + e) / 2);
    int aS = A.get(s), aE = A.get(e), aM = A.get(m);
    int pivot = s; 
    if(aM > aS && aM < aE) {
      pivot = m;
    } else if(aE > aS && aE < aM) {
      pivot = e;
    }
    int temp = A.get(s);
    A.set(s, A.get(pivot));
    A.set(pivot, temp);
    int l = s + 1, r = e;
    while(l <= r) {
      if(A.get(l) <= A.get(s)) {
        l++;
      } else if(A.get(r) >= A.get(s)) {
        r--;
      } else {
        temp = A.get(l);
        A.set(l, A.get(r));
        A.set(r, temp);
        l++;
        r--;
      }
    }
    temp = A.get(s);
    A.set(s, A.get(r));
    A.set(r, temp);

    return r;
  }

  /**
   * Q3. Quick Sort
   *
   * Given an integer array A, sort the array using Quick Sort.
   */
  private static void quickSort(ArrayList<Integer> A, int s, int e) {
    if(s >= e) {
      return;
    }

    int p = parition(A, s, e);
    quickSort(A, s, p - 1);
    quickSort(A, p + 1, e);
  }  

  public static void main(String[] args) {
    ArrayList<Integer> A = new ArrayList<>(Arrays.asList(4,5,6,3,1,45,6,3,5,6));
    quickSort(A, 0, A.size() - 1);
    System.out.println(A);
  }
}
